src/ddmd/irstate.d

```diff
@@ -76,13 +77,19 @@ struct IRState
             deferToObj = irs.deferToObj;
             varsInScope = irs.varsInScope;
             labels = irs.labels;
+            mayThrow = irs.mayThrow;
         }
     }
 
-    this(Module m, FuncDeclaration s)
+    this(Module m, FuncDeclaration fd, Array!(elem*)* varsInScope, Dsymbols* deferToObj, Label*[void*]* labels)
     {
         this.m = m;
-        symbol = s;
+        this.symbol = fd;
+        this.varsInScope = varsInScope;
```

Comment:
I see that the assumption in the rest of the code is that `irs.varsInScope` can't be `null`. How about changing the constructor parameter to `ref Array!(elem*) varsInScope` to solidify this assumption?

Reply:
Because then I have to take the address of a `ref` parameter which is not allowed in `@safe` code.

Reply:
Besides, if it is null, it'll seg fault on first use. So I'm not worried about it :-)

Reply:
> Besides, if it is null, it'll seg fault on first use. So I'm not worried about it getting through the test suite with such a bug :-)

I'm not that worried about bugs, as much I would like to make the interface more clear for the callers.

> Because then I have to take the address of a ref parameter which is not allowed in @safe code.

Taking the address of a local variable and passing it to a non-`scope` parameter is not `@safe` either. Ideally (in the distant future) the ownership could be clarified, so even the the compiler could see it is `@safe`:
```D
// in some function
{
    Array!(elem*) varsInScope;
    {
        // it's safe to pass `varsInScope` as `lifetime(varsInScope) > lifetime(irs)`
        IRState irs = IRState(.., /* scope ref parameter */ varsInScope, ..);
    }
}
```


Anyway, it's just a suggestion - we can revisit the discussion some other day if you want to move forward with the code as it is.